sklearn/metrics/_pairwise_distances_reduction/_base.pyx

```diff
@@ -97,7 +97,7 @@ cdef class PairwiseDistancesReduction64:
         if strategy == 'auto':
             # This is a simple heuristic whose constant for the
             # comparison has been chosen based on experiments.
-            if 4 * self.chunk_size * self.effective_n_threads < self.n_samples_X:
+            if self.n_samples_Y < self.n_samples_X:
                 strategy = 'parallel_on_X'
             else:
                 strategy = 'parallel_on_Y'
```

Comment:
Yes having both choices is currently better, but probably for wrong reason: I noted that the choice of the first one can be improved for:

```python
            if C * self.n_samples_Y < self.n_samples_X:
```

where at first sight `C in [1, 10]`.
This makes sense, because `parallel_on_Y` is more costly than `parallel_on_X` for datasets of the same size (because there are extra data-structures and non-blocking synchronisation for `parallel_on_Y`). This way we can probably come up with a simpler heuristic? I can't remember why we chose the second option in the first place (to me it is not a valid one).

Reply:
I have the feeling that when both X is smaller than Y and is large, parallel on X might still be advantageous or not depending on the number of available hardware threads.

Since parallel on X has less synchronization overhead, I would expect it to be always advantageous as long as each thread processes a few chunks of X (e.g. more than 2 to 10), even if Y 100 times larger than X.

Hence my suggestion above. Off-course this should be confirmed by re-running the benchmarks.